R/fabricate.R

```diff
@@ -35,215 +52,667 @@
+        )
+    }
+  }
 
-  if (length(options) == 0)  return(FALSE)
+  # There are two possibilities. One is that we are modifying the lowest level
+  # of data. In which case, we simply add variables, like if someone called
+  # add_level with a dataset. To check if that's the world we're in, check if
+  # we have any duplicates in the ID label:
+  if(!anyDuplicated(working_environment_$data_frame_output_[[ID_label]])) {
+    # There is no subsetting going on, but modify_level was used anyway.
+    N = nrow(working_environment_$data_frame_output_)
+
+    # Coerce the working data frame into a list
+    working_data_list = as.list(working_environment_$data_frame_output_)
+
+    # Now loop over the variable creation.
+    for(i in names(data_arguments)) {
+      # Evaluate the formula in an environment consisting of:
+      # 1) The current working data list
+      # 2) A list that tells everyone what N means in this context.
+      # Store it in the current environment
+      working_data_list[[i]] = eval_tidy(data_arguments[[i]],
+                                         append(working_data_list, list(N=N)))
+
+      # Write the variable name to the list of variable names
+      add_variable_name(working_environment_, i)
+
+      # Nuke the current data argument -- if we have the same variable name
+      # created twice, this is OK, because it'll only nuke the current one.
+      data_arguments[[i]] = NULL
+    }
 
-  is_function <- sapply(options, function(i) {
-      is_lang(get_expr(i))
-  })
+    # Before handing back data, ensure it's actually rectangular
+    working_data_list = check_rectangular(working_data_list, N)
 
-  is_level <- "level" == sapply(options[is_function], lang_name) ## function names
+    # Overwrite the working data frame.
+    working_environment_[["data_frame_output_"]] = data.frame(working_data_list,
+                                                              stringsAsFactors=FALSE,
+                                                              row.names=NULL)
 
-  if(length(is_level) == 0) return(FALSE)
+    # Return results
+    return(working_environment_)
+  }
 
-  if (any(is_level) != all(is_level)) {
+  # If we're here, then at least some subsetting is used in the modify call
+  # first, subset to unique observations, then generate new data, then re-expand.
+  # To do this, we need a mapping between observations and unique observations.
+  # First, get the unique values of the level:
+  unique_values_of_level = unique(working_environment_$data_frame_output_[[ID_label]])
+
+  # Pre-allocate the mapping vector
+  index_maps = numeric(length(working_environment_$data_frame_output_[[ID_label]]))
+  # Iterate along the unique values of the level
+  for(i in seq_along(unique_values_of_level)) {
+    # Any obs that matches the level matching this i will be a duplicate of this i.
+    index_maps[
+      working_environment_$data_frame_output_[[ID_label]] == unique_values_of_level[i]
+      ] = i
```

Comment:
yikes.

Reply:
Yeah I'm really open to a better implementation of this, I'm sure there is one.